### DISCUSSION TITLE
Allow separate images for 100/50 that are early versus late

### DISCUSSION THREAD

**Discussion by DanielPower** [ppy/osu#22603] [Ideas]
Currently skins can provide an image that should be shown for 100 points and 50 points. When those images are shown in-game it gives visual feedback that your timing is off, but it does not tell you how your timing is off.

Averting your eyes away from the notes to look at the error bar is distracting and can result in mis-reading.

I'd like to propose the ability for skins to provide separate images for 100-early, 100-late, 50-early, 50-late. This way skins could distinguish between early and late, perhaps by using different colors, so that players can visually distinguish whether they need to speed up or slow down.

**Comment by mk56-spn** [comment]
https://github.com/ppy/osu/discussions?discussions_q=early+late

**Reply by DanielPower** [reply]
If you're going to claim my idea is a duplicate, could you link to the actual duplicate discussion, instead of a search query containing a few dozen unrelated discussions?

If you're implying I didn't search before posting this, you're wrong. Though I didn't use the exact "early+late" query you posted. Now, using that I've found this discussion. https://github.com/ppy/osu/discussions/14858 Which Peppy responded to with an equally unhelpful link to a search query, instead of an actual discussion.

**Reply by DanielPower** [reply]
https://github.com/ppy/osu/discussions/14038 There's this discussion, which is worded strangely, but _could_ be asking for the same thing I am? I'm not sure though. The way it's worded makes it seem like they're asking for customization of scoring functionality, not skinning.

**Reply by mk56-spn** [reply]
It will come at some point. When that'll be I cant really say

**Reply by bdach** [reply]
The actual duplicate here is https://github.com/ppy/osu/issues/3209

**Reply by DanielPower** [reply]
Thank you! It looks like Peppy responded positively to that issue. I've been interested in working on Lazer's code, so I'll take a swing at it :)

**Reply by bdach** [reply]
Sorry but I'm not sure it's a good idea, at least not right now. I'm not sure how this would even be implemented.

I'm pretty sure we have made the decision to not extend legacy (stable) skins with new features.

**Reply by DanielPower** [reply]
I was thinking I could attempt to implement it under the new skinning architecture for Argon. I understand that legacy skins are not receiving new functionality. Legacy skins would maintain their current functionality. I'm not sure this kind of feature would be feasible as a first task, but I'm interested to investigate it.

It looks like the skin doesn't know the timing information, it's just given a member from the `HitResult` enum, in order to determine which judgement to display.

One way to go about implementing this could be to introduce new HitResult members, such that there is an early and late member for each judgement.

ok_early
ok_late
good_early
good_late
etc.

Argon and new skins could then display something different for early versus late. Whereas legacy skins would display the same image for both, maintaining previous functionality.

Although I don't have the experience with the codebase to know how much of a rabbit hole I'd be digging by introducing new hit results, especially when factoring in game modes other than standard, and a skinning system I don't fully understand yet.

**Reply by bdach** [reply]
Adding new `HitResult`s is incorrect. The correct way to proceed would involve reading the hit offset from the judgement.
